### PR TITLE
[FLINK-7449] [docs] Additional documentation for incremental checkpoints

### PR DESCRIPTION
**Pull request by StefanRRichter**
## What is the purpose of the change

This PR provides additional documentation for incremental checkpoints.

## Brief change log

Added documentation.

## Does this pull request potentially affect one of the following parts:

No

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

**Comment by StefanRRichter** [comment]
CC @twalthr

File: docs/ops/state/checkpoints.md [hunk 1]
```diff
@@ -99,3 +99,296 @@ above).
...
+(hopefully few) *events that have been processed after the last completed checkpoint* must be reprocessed from the data
+source. The number of events that must be reprocessed has implications for recovery time, and so for fastest recovery,
+we want to *take checkpoints as often as possible*.
+
+However, checkpoints are not without performance cost and can introduce *considerable overhead* to the system. This
+overhead can lead to lower throughput and higher latency during the time that checkpoints are created. One reason is
+that, traditionally, each checkpoint in Flink always represented the *complete state* of the job at the time of the
+checkpoint, and all of the state had to be written to stable storage (typically some distributed file system) for every
+single checkpoint. Writing multiple terabytes (or more) of state data for each checkpoint can obviously create
+significant load for the I/O and network subsystems, on top of the normal load from pipeline’s data processing work.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: on top of the normal load from the pipeline’s data processing work.

(add "the")

File: docs/ops/state/checkpoints.md [hunk 2]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+At this point, it looks like we didn’t gain too much from incremental checkpoints because we are again trading between
+checkpointing overhead and recovery time. Fortunately, there are ways to improve on this naive approach to recovery. One
+simple and obvious way to restrict recovery time and the length of the checkpoint history is to write a full checkpoint
+from time to time. We can drop all checkpoints prior to the most recent full checkpoint, and the full checkpoint can
+serve as a new basis for future incremental checkpoints.
+
+Our actual implementation of incremental checkpoints in Flink is more involved and designed to address a number of
+different tradeoffs. Our incremental checkpointing restricts the size of the checkpoint history and therefore never
+needs take a full checkpoint to keep recovery efficiently! We present more detail about this in the next section, but
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: to keep recovery efficient!

File: docs/ops/state/checkpoints.md [hunk 3]
```diff
@@ -99,3 +99,296 @@ above).
...
+the high level idea is to accept a small amount of redundant state writing to incrementally introduce
+merged/consolidated replacements for previous checkpoints. For now, you can think about Flink’s approach as stretching
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: the high level idea is to accept a small amount of redundant state writing that incrementally introduces
merged/consolidated replacements for previous checkpoints.

File: docs/ops/state/checkpoints.md [hunk 4]
```diff
@@ -99,3 +99,296 @@ above).
...
+out and distributing the consolidation work over several incremental checkpoints, instead of doing it all at once in a
+full checkpoint. Every incremental checkpoint can contribute a share for consolidation. We also track when old
+checkpoints data becomes obsolete and then prune the checkpoint history over time.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: when old checkpoint data OR when previously checkpointed data

File: docs/ops/state/checkpoints.md [hunk 5]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+### Incremental Checkpoints in Flink
+
+In the previous section, we discussed that incremental checkpointing is mainly about recording all effective state
+modifications between checkpoints. This poses certain requirements on the underlying data structures in the state
+backend that manages the job’s state. It goes without saying that the data structure should always provide a decent
+read-write performance to keep state access swift. At the same time, for incremental checkpointing, the state backend
+must be able to efficiently detect and iterate state modifications since the previous checkpoint.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: and iterate over state modifications

File: docs/ops/state/checkpoints.md [hunk 6]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+One data structure that is very well-suited for this use case is the *log-structured-merge (LSM) tree* that is the core
+data structure in Flink’s RocksDB-based state backend. Without going into too much detail, the write path of RocksDB
+already roughly resembles a changelog with some pre-aggregation — which perfectly fits the needs of incremental
+checkpoints. On top of that, RocksDB also has a *compaction mechanism* can be regarded as an elaborated form of log
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: which can be regarded

File: docs/ops/state/checkpoints.md [hunk 7]
```diff
@@ -99,3 +99,296 @@ above).
...
+compaction.
+
+#### RocksDB Snapshots as a Foundation
+
+In a nutshell, *RocksDB is a key-value store based on LSM trees*. The write path of RocksDB first collects all changes as
+key-value pairs in a mutable, in-memory buffer called memtable. Writes to the same key in a memtable can simply replace
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: called a memtable. ... Once a memtable is full, ... After a memtable is written to disk, it becomes immutable and is then called a ...

File: docs/ops/state/checkpoints.md [hunk 8]
```diff
@@ -99,3 +99,296 @@ above).
...
+(sstable)*. Figure 2 illustrates these basic RocksDB internals.
+
+<p class="text-center">
+   <img alt="Figure 2: RocksDB architecture (simplified)" width="75%" src="{{ site.baseurl }}/fig/rocksdb_architecture_simple.png"/>
+</p>
+
+To avoid the problem of collecting an infinite number of sstables over time, a background task called compaction is
+constantly merging sstables to consolidate potential duplicate entries for each key from the merged tables. Once some
+sstables have been merged, those original sstables become obsolete and are deleted by RocksDB. The newly created merged
+sstable contains all their net information. We show an example for such a merge in Figure 3. SSTable-1 and SStable-2
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: contains all of their net information.

File: docs/ops/state/checkpoints.md [hunk 9]
```diff
@@ -99,3 +99,296 @@ above).
...
+that deleted sstables in RocksDB are always the result of compaction. This is the way in which Flink’s incremental
+checkpoints can prune the checkpoint history. Old sstables are eventually replaced by the sstable that is the result of
+merging them. Note that in a strict sense of tracking changes between checkpoints, this uploading of consolidated tables
+is redundant work. But it is performed incrementally, typically adding only a small amount of overhead to some
+checkpoints. However, we absolutely consider that overhead to be a worthwhile investment because it allows us to keep a
+shorter history of checkpoints to consider in a recovery.
+
+Another interesting point is how Flink can determine when it is safe to delete a shared file. Our solution works as
+follows: for each file, we keep a reference count for each sstable file that we copied to stable storage. These counts
+are maintained by the checkpoint coordinator on the job master in a *shared state registry*. This shared registry tracks
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: job manager

File: docs/ops/state/checkpoints.md [hunk 10]
```diff
@@ -99,3 +99,296 @@ above).
...
+completed. For checkpoint 1 (``CP 1``)), we can see that the local RocksDB directory contains two sstable files, which
+are considered as new and uploaded to stable storage. We upload the files under the checkpoint directory of the
+corresponding checkpoint that first uploaded them, in this case ``cp-1``, and use unique filenames because they could
+otherwise collide with identical sstable names from other subtasks. When the checkpoint completes, the two entries are
+created in the shared state registry, one for each newly uploaded file, and their counts are set to 1. Notice that the
+key in the shared state registry is a composite of operator, subtask, and the original sstable file name. In the actual
+implementation, the shared state registry also keeps a mapping from the key to the file path in stable storage besides
+the count, which is not shown to keep the graphic clearer.
+
+At the time of the second checkpoint, two new sstable files have been created by RocksDB and the two older sstable files
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: two new sstable files have been created by RocksDB,

File: docs/ops/state/checkpoints.md [hunk 11]
```diff
@@ -99,3 +99,296 @@ above).
...
+from the previous checkpoint also still exist. For checkpoint ``CP 2``, Flink must now upload the two new files to
+stable storage and can reference the ``sstable-(1)`` and ``sstable-(2)`` from the previous checkpoint. We can see that
+the file references to previously existing sstable files point to existing files in the ``cp-1`` directory and
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: point to existing files in the ``cp-1`` directory,

File: docs/ops/state/checkpoints.md [hunk 12]
```diff
@@ -99,3 +99,296 @@ above).
...
+from a previous checkpoint. When this checkpoint completes, two things will happen at the checkpoint coordinator, in the
+following order:
+
+* First, the checkpoint registers the referenced files, increasing the count of those files by 1.
+
+* Then, the older checkpoint ``CP 1`` will be deleted because we have configured the number of retained checkpoints to
+two.
+
+* As part of the deletion, the counts for all files referenced by ``CP 1``, (``sstable-(1)`` and ``sstable-(2)``), is
+decreased by 1.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: the counts ... are decreased by 1

File: docs/ops/state/checkpoints.md [hunk 13]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+We sometimes also have to resolve race conditions between concurrent checkpoints in incremental checkpointing. Flink can
+execute multiple checkpoints in parallel, and new checkpoints can start before previous checkpoints are confirmed as
+completed by the checkpoint coordinator to the backend. We need to consider this in our reasoning about which previous
+checkpoint can serve as a basis for a new incremental checkpoint. We are only allowed to reference shared state from a
+confirmed checkpoint, because otherwise we might attempt to reference a shared file that might still be deleted, e.g.
+when the assumed predecessor checkpoint still fails.
+
+This can lead to a situation were multiple checkpoints regard the same sstable files in RocksDB as new because no
+checkpoint that attempted to upload and register those sstable files has been confirmed, yet. To be on the safe side,
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: has yet been confirmed.

File: docs/ops/state/checkpoints.md [hunk 14]
```diff
@@ -99,3 +99,296 @@ above).
...
+checkpoints must always upload such files to stable storage independently, under unique names, until the sstable files
+have been registered by a completed checkpoint and the confirmation reached the backend. Otherwise, pending previous
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: and the confirmation has reached the backend.

File: docs/ops/state/checkpoints.md [hunk 15]
```diff
@@ -99,3 +99,296 @@ above).
...
+checkpoints might still fail, in which case their newly uploaded files are deleted, and future checkpoints would
+potentially attempt to reference deleted data.
+
+Sometimes, this upload policy will result in the same sstable file been uploaded more than once, from different
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: This upload policy can result in the same sstable file being uploaded more than once, from different checkpoints.

File: docs/ops/state/checkpoints.md [hunk 16]
```diff
@@ -99,3 +99,296 @@ above).
...
+checkpoints. However, at least we can later de-duplicate the sstable files in the checkpoint coordinator because they
+are accounted under the same key. Only the copy that was uploaded by the first-confirmed checkpoint survives and we can
+replace reference to the duplicates in all checkpoints that register afterwards.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: Only the copy that was uploaded by the first-confirmed checkpoint survives, and we can
replace references to the duplicates in all checkpoints that register afterwards.

File: docs/ops/state/checkpoints.md [hunk 17]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+Incremental checkpoints are only available for checkpoints and not for savepoints. Savepoints are always self-contained
+and record the full state of a job. However, it is possible to externalize incremental checkpoints. This is a way to use
+them for manual restarts of a job.
+
+Rescaling the parallelism of a job is an operation that is officially only supported by through savepoints and not from
+incremental checkpoints. (Unofficially, it should still be possible, though.)
+
+Users should not manually delete non-empty checkpoint directories when working with incremental checkpoints. A newer
+checkpoint might still reference files from the doctor of an older checkpoint.
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: doctor -> directory

File: docs/ops/state/checkpoints.md [hunk 18]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+Very small checkpoint intervals can become another problem in connection with incremental checkpoints. As we have
+previously explained, the first step of an incremental checkpoint is forcing RocksDB to flush all memtables to disk. If
+this step is performed very frequently, it can created a lot of small files. Too many small files can become problematic
```
Comment (alpinegizmo) [inline docs/ops/state/checkpoints.md]: can create

**Comment by ChrisChinchilla** [comment]
@StefanRRichter Also, some of the images have 'spelling errors' highlighted in them.

File: docs/ops/state/checkpoints.md [hunk 19]
```diff
@@ -99,3 +99,296 @@ above).
...
+
+### Basics of Incremental Checkpoints
+
+In this section, for the sake of getting the concept across, we will briefly discuss the idea behind incremental
+checkpoints in a simplified manner.
+
+Our motivation for incremental checkpointing stemmed from the observation that it is often wasteful to write the full
+state of a job for every single checkpoint. In most cases, the state between two checkpoints is not drastically
+different, and only a fraction of the state data is modified and some new data added. Instead of writing the full state
+into each checkpoint again and again, we could record only changes in state since the previous checkpoint. As long as we
```
Comment (ChrisChinchilla) [inline docs/ops/state/checkpoints.md]: @StefanRRichter I'm struggling to understand this sentence.

> As long as we have the previous checkpoint and the state changes for the current checkpoint, we can restore the full, current state for the job.

Do you mean…

> As long as we have the previous checkpoint, if the state changes for the current checkpoint, we can restore the full, current state for the job.

Or something different? Explain to me what you're trying to say :)

File: docs/ops/state/checkpoints.md (hunk 2 above)
Comment (ChrisChinchilla) [inline docs/ops/state/checkpoints.md]: @StefanRRichter Is the whole section…

> We present more detail about this in the next section…

Until the end of the paragraph needed? If you're about to cover it below anyway, why mention it? Or is there anything here that isn't mentioned in the 

> ### Incremental Checkpoints in Flink

section?

**Comment by tillrohrmann** [comment]
Ping @StefanRRichter. Can we get this documentation merged? The community would definitely appreciate it.

**Pull request closed without merging**